A43-sdk-authorization.md

```diff
@@ -0,0 +1,364 @@
+			"source": {
+				"principals": [
+					"*"
+				]
+			},
+			"request": {
+				"paths": [
+					"/pkg.service/foo",
+					"/pkg.service/bar"
+				]
+			}
+		}
+	],
+	"deny_rules": [
+		{
+			"name": "deny-access",
+			"request": {
+				"paths": [
+					"*/secret"
+				]
+			}
+		}
+	]
+}
+```
+
+As mentioned previously, internally gRPC Authorization engine consumes Envoy RBAC 
+proto. The user provided JSON SDK authorization policy will be translated to Envoy 
+RBAC proto and provided to Engines for evaluation. Note that SDK authorization policy 
+is a subset of Envoy RBAC, and it does not support all the fields that are present in 
+Envoy RBAC.
+
+### API
+
+gRPC will support both static initialization and dynamically reloading the policy
+from filesystem. In static initialization, the policy will be provided as a JSON
+string. In dynamic file reloading, the application will specify the file path that
+contains the authorization policy in JSON format.
+
+Following code snippets show how to enable authorization in gRPC servers in 
+different languages.
+
+#### C++
+
+Static Initialization
+
+```C++
+std::shared_ptr<AuthorizationPolicyProviderInterface> provider = 
+	StaticDataAuthorizationPolicyProvider::Create(authz_policy);
+ServerBuilder builder;
+builder.SetAuthorizationPolicyProvider(provider);
+std::unique_ptr<grpc::Server> server = builder.BuildAndStart();
+```
+
+Dynamic file reloading
+
+```C++
+std::shared_ptr<AuthorizationPolicyProviderInterface> provider = 
+	FileWatcherAuthorizationPolicyProvider::Create(authz_policy_path, /*refresh_interval_sec=*/3600);
+ServerBuilder builder;
+builder.SetAuthorizationPolicyProvider(provider);
+std::unique_ptr<grpc::Server> server = builder.BuildAndStart();
+```
+
+#### Go
+
+Static initialization
+
+```Go
+creds := credentials.NewServerTLSFromFile(certFile, keyFile)
+i, err := authz.NewStatic(authzPolicy)
+serverOpts := []grpc.ServerOption{
+  grpc.Creds(creds),
+  grpc.ChainUnaryInterceptor(i.UnaryInterceptor),
+  grpc.ChainStreamInterceptor(i.StreamInterceptor),
+}
+s := grpc.NewServer(serverOpts...)
+```
+
+Dynamic File Reloading
+
+```Go
+creds := credentials.NewServerTLSFromFile(certFile, keyFile)
+i, err := authz.NewFileWatcher(authzPolicyFile, 1 * time.Hour)
+serverOpts := []grpc.ServerOption{
+  grpc.Creds(creds),
+  grpc.ChainUnaryInterceptor(i.UnaryInterceptor),
+  grpc.ChainStreamInterceptor(i.StreamInterceptor),
+}
+s := grpc.NewServer(serverOpts...)
+// In the end, free up the resources used for policy refresh.
+defer i.Close()
+```
+
+#### Java
+
+Static initialization
+
+```Java
+AuthorizationServerInterceptor authzServerInterceptor;
+try {
+  authzServerInterceptor = AuthorizationServerInterceptor.create(authzPolicy);
+} catch (IllegalArgumentException iae) {
+  // ...
+}
+Server server =
+    ServerBuilder.forPort(port)
+        .addService(service)
+        .useTransportSecurity(certChainFile, privateKeyFile)
+        .intercept(authzServerInterceptor);
```

Comment:
1) What happens if you provide multiple AuthorizationServerInterceptor interceptors this way? How will you catch such a situation?
2) Is it possible to have a simpler interface similar to C++ so the user doesn't know it's an interceptor e.g.
```
AuthorizationPolicyProvider authzPolicyProvider = AuthorizationPolicyProvider.create(authzPolicy);
Server server =    ServerBuilder.forPort(port)
        .addService(service)
        .useTransportSecurity(certChainFile, privateKeyFile)
        .authzPolicyProvider(authzPolicyProvider);
        .build().start();
```

Reply:
That is a very good question. With the current approach, we could provide multiple AuthorizationServerInterceptors and each RPC will be validated against these interceptors.
We can avoid that and limit the SDK interceptor to just one instance using the wrapper approach u mentioned. And it looks much cleaner. 

Is there a reason why we chose to use interceptors directly in the Java API design? @ejona86 @sanjaypujare

Reply:
Based on email discussion, we have decided to keep the current design.